cairo/ethereum/cancun/vm/instructions/log.cairo

```diff
@@ -0,0 +1,445 @@
+    if (cast(err, felt) != 0) {
+        return err;
+    }
+
+    // OPERATION
+    let memory = evm.value.memory;
+    with memory {
+        expand_by(extend_memory.value.expand_by);
+        let data = memory_read_bytes(memory_start_index, size);
+    }
+
+    // Create log entry
+    let null_ptr = cast(0, Bytes32Struct*);
+    tempvar topics = TupleBytes32(new TupleBytes32Struct(new Bytes32(null_ptr), 0));
+    tempvar log_entry = new Log(
+        new LogStruct(address=evm.value.message.value.current_target, topics=topics, data=data)
+    );
+
+    // Check for static context
+    if (evm.value.message.value.is_static.value != 0) {
+        tempvar err = new ExceptionalHalt(WriteInStaticContext);
+        return err;
+    }
+
+    // Append log entry
+    EvmImpl.set_logs(TupleLog(new TupleLogStruct(log_entry, 1)));
+
+    // PROGRAM COUNTER
+    EvmImpl.set_pc_stack_memory(Uint(evm.value.pc.value + 1), stack, memory);
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice LOG1 instruction - append log record with one topic
+func log1{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, evm: Evm}() -> ExceptionalHalt* {
+    alloc_locals;
+    const NUM_TOPICS = 1;
+
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        // Pop memory_start_index, size and topic
+        let (memory_start_index, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+        let (size, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+        let (topic, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+    }
+
+    // GAS
+    // Calculate memory expansion cost
+    // If the size is greater than 2**128, the memory expansion will trigger an out of gas error.
+    if (size.value.high != 0) {
+        tempvar err = new ExceptionalHalt(OutOfGasError);
+        return err;
+    }
+
+    // Calculate memory expansion cost
+    tempvar mem_access_tuple = new TupleU256U256(new TupleU256U256Struct(memory_start_index, size));
+    tempvar mem_access_list = ListTupleU256U256(new ListTupleU256U256Struct(mem_access_tuple, 1));
+    let extend_memory = calculate_gas_extend_memory(evm.value.memory, mem_access_list);
+
+    // Calculate total gas cost
+    let data_cost = GasConstants.GAS_LOG_DATA * size.value.low;
+    let topic_cost = GasConstants.GAS_LOG_TOPIC * NUM_TOPICS;
+    let total_cost = Uint(
+        GasConstants.GAS_LOG + data_cost + topic_cost + extend_memory.value.cost.value
+    );
+
+    let err = charge_gas(total_cost);
+    if (cast(err, felt) != 0) {
+        return err;
+    }
+
+    // OPERATION
+    let memory = evm.value.memory;
+    with memory {
+        expand_by(extend_memory.value.expand_by);
+        let data = memory_read_bytes(memory_start_index, size);
+    }
+
+    // Create log entry
+    let be_topic_ = U256_to_be_bytes(topic);
+    tempvar be_topic = new Bytes32(be_topic_.value);
```

Comment:
```suggestion
    tempvar be_topic = Bytes32(be_topic_.value);
```

Reply:
if I remove the `new` keyword, then in:
`tempvar topics = TupleBytes32(new TupleBytes32Struct(&be_topic, 1));`
I have to add the `&` pointer to construct `TupleBytes32Struct`.
but if I do so, I get:
```
tempvar topics = TupleBytes32(new TupleBytes32Struct(&be_topic, 1));
INTERNALERROR>                                                           ^******^
INTERNALERROR> :152:13: ap may only be used in an expression of the form [ap + <const>].
INTERNALERROR>     tempvar be_topic = Bytes32(be_topic_.value);
INTERNALERROR>             ^******^
```